### PR TITLE
Set unit names according to their faction

### PR DESCRIPTION
**Pull request by MadBender**
## What type of PR is this.
1. [ ] Bug
2. [ ] Change
3. [x] Enhancement

### What have you changed and why?
All units used to have Greek names regardless of their faction. Now unit names depend on their faction. NATO units have Western names, CSAT will have Chinese/Iranian names etc. Names were customized for vanilla, RHS and CUP factions, other faction will still use Greek names, this can be fixed later.

`A3A_fnc_createUnit` now always sets the unit's identity to ensure that each unit has his identity configured. There was a bug when faction identity did not apply to rebel garrisons and maybe in some other cases.

### Please verify the following and ensure all checks are completed.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [x] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: 

Find a field, look to the north and run
```sqf
private _pos = getPos player;

private _fnc_makeSafe = {
    params ["_group"];
    _group setCombatMode "BLUE"; 
    _group setBehaviourStrong "CARELESS"; 
    {
        _x disableAI "TARGET";  
        _x disableAI "AUTOTARGET"; 
        _x stop true; 
        _x setCaptive true; 
    } forEach units _group;
};

private _fnc_getSquad = {
    params ["_type"];
    (_type splitString "_") select 0;
};

private _fnc_spawnFaction = {
    params ["_faction", "_side", "_prefix", "_pos"];
    private _group = createGroup _side;
    private _loadouts = keys(_faction get "loadouts");
    _loadouts sort true;
    private _lastSquad = "";
    {
        private _squad = _x call _fnc_getSquad;
        private _offset = switch (_lastSquad) do {
            case (""): {0};
            case (_squad): {2};
            default {4};
        };
        _pos = _pos vectorAdd [_offset, 0, 0];

        private _loadout = format["loadouts_%1_%2", _prefix, _x];
        [_group, _loadout, _pos] call A3A_fnc_createUnit;
        _lastSquad = _squad;
    } forEach _loadouts;
    _group call _fnc_makeSafe;
    _group;
};

private _rebels = [A3A_faction_reb, resistance, "reb", _pos vectorAdd[0, 5, 0]] call _fnc_spawnFaction;
{
    [_x, _x getVariable "unitType"] call A3A_fnc_FIAinit;
} forEach units _rebels;

private _occupants = [A3A_faction_occ, west, "occ", _pos vectorAdd[0, 10, 0]] call _fnc_spawnFaction;
{
    _x call A3A_fnc_NATOinit;
} forEach units _occupants;

private _invaders = [A3A_faction_inv, east, "inv", _pos vectorAdd[0, 15, 0]] call _fnc_spawnFaction;
{
    _x call A3A_fnc_NATOinit;
} forEach units _invaders;

[A3A_faction_civ, civilian, "civ", _pos vectorAdd[0, 20, 0]] call _fnc_spawnFaction;
```

### CODE REVIEW

**Comment by MadBender** [comment]
Names are taken from `configfile >> "CfgWorlds" >> "GenericNames"`

image

Normally, these names are referenced via `genericNames` parameter in the unit class, e.g. `configfile >> "CfgVehicles" >> "B_Soldier_A_F" >> "genericNames"`. I placed different units in the editor and checked their `genericNames` and put what I found into the faction templates.

Mods can declare their own name lists or use vanilla lists, for example all US-based factions use `NATOMen` from vanilla. `TakistaniMen` is also from vanilla.

**Review by Lazejun — CHANGES_REQUESTED** [review]
Why did you unmute the Civs?

Your Whole PR talks about Setting names and changing a few Functions, while you are silently unmuting Civs.

**Comment by MadBender** [comment]
@Lazejun since civilian factions don't have `voices` defined, `createRandomIdentity` will set `NoVoice` for them, and they will remain muted

https://github.com/MadBender/A3-Antistasi/blob/e16c1f9fb938f25e317133a7ce18affaf1ce9b2d/A3A/addons/core/functions/Utility/fn_createRandomIdentity.sqf#L39-L45

**Review by Lazejun — CHANGES_REQUESTED** [review]
"this can be fixed later."

Well yes, if you add a Placeholder.

https://github.com/official-antistasi-community/A3-Antistasi/blob/355b5676d112bd8b853bcd8a3ae2f4f23a05c4a1/A3A/addons/core/Templates/Templates/FactionDefaults/EnemyDefaults.sqf#L14

```sqf
"GreekMen" call _fnc_saveNames;
```

else you are gonna get for each faction thats not done yet: 

```
12:57:30 Error in expression <tity getOrDefault ["lastName", ""];
if (_firstName != "" || _lastName != "") the>
12:57:30   Error position: <_firstName != "" || _lastName != "") the>
12:57:30   Error Undefined variable in expression: _firstname
12:57:30 File \x\A3A\addons\core\functions\Utility\fn_setIdentityLocal.sqf [A3A_fnc_setIdentityLocal]..., line 29
12:57:30  ➥ Context: 	[] L29 (\x\A3A\addons\core\functions\Utility\fn_setIdentityLocal.sqf [A3A_fnc_setIdentityLocal])
```

**Comment by MadBender** [comment]
set default voices, faces and names for all factions, so the code may assume they are always present

**Review by Lazejun — APPROVED** [review]

**Review by jaj22 — APPROVED** [review]
Code looks good and moving the identity-selection code to createUnit is an improvement. My one caveat is that you're passing more data over the network unnecessarily with the hashmap-parameters. It's probably negligible compared to other costs of creating units (esp. with our template system) but Arma's so inefficient at passing SQF data over the network that it makes me nervous.

On balance I think merge as-is and maybe optimize the parameter passing later if I get the urge. Should probably have been in the 3.3 test but that caught me a bit cold.

File: A3A/addons/core/functions/Base/fn_createPetros.sqf [hunk 1]
```diff
@@ -16,7 +16,13 @@ if (isNil "_location") then {
 };
 
 private _oldPetros = petros;
-petros = [_groupPetros, FactionGet(reb,"unitPetros"), _location, [], 10, "NONE"] call A3A_fnc_createUnit;
+private _identity = createHashMapFromArray[
```
Comment (rautamiekka) [inline A3A/addons/core/functions/Base/fn_createPetros.sqf:19]:
```suggestion
private _identity = createHashMapFromArray [
```

File: A3A/addons/core/functions/Utility/fn_createRandomIdentity.sqf [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+Scope: Any
+Environment: Any
+Public: Yes
+
+Example:
+    private _identity = [A3A_faction_reb, "Rifleman"] call A3A_fnc_createRandomIdentity;
+*/
+
+params["_faction", "_unitType"];
```
Comment (rautamiekka) [inline A3A/addons/core/functions/Utility/fn_createRandomIdentity.sqf:24]:
```suggestion
params ["_faction", "_unitType"];
```

File: A3A/addons/core/functions/init/fn_initClient.sqf [hunk 3]
```diff
@@ -122,7 +122,7 @@ stragglers = creategroup teamPlayer;
 (group player) enableAttack false;
 
 if (isNil "ace_noradio_enabled" or {!ace_noradio_enabled}) then {
-    [player, nil, selectRandom (A3A_faction_reb get "voices")] call A3A_fnc_setIdentity
+    [player, createHashMapFromArray[["speaker", selectRandom (A3A_faction_reb get "voices")]]] call A3A_fnc_setIdentity
```
Comment (rautamiekka) [inline A3A/addons/core/functions/init/fn_initClient.sqf:125]:
```suggestion
    [player, createHashMapFromArray [["speaker", selectRandom (A3A_faction_reb get "voices")]]] call A3A_fnc_setIdentity
```

**Review by rautamiekka — CHANGES_REQUESTED** [review]

**Comment by jaj22** [comment]
Probably needs updating for the new templates (RHS chdkz, various SPE stuff, did I miss anything?). If you don't have SPE then I'll deal with it.

**Comment by Bob-Murphy** [comment]
New templates need to be adapted as well.

**Comment by MadBender** [comment]
Sorry for delay, don't have much time for ARMA right now.

I never intended to configure all factions in a single PR, I don't even have most of paid DLCs. So I made this configuration optional. Unconfigured factions will work as before and can be configured later.

If this is the only problem then I suggest to merge to unstable and test with many players. We can make small PRs for other factions later

**Comment by jaj22** [comment]
Yeah checked, it uses the template defaults to set greek voice/face/name, so it shouldn't break anything badly on new templates and should be good for merge.

**Pull request merged**